docs/site/migration/models/mixins.md

```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+    async findByTitle(title: string): Promise<M[]> {
+      const titleFilter = {
+        where: {
+          title: title,
+        },
+      };
+      const foundItems = await this.find(titleFilter);
+      return foundItems;
+    }
+  };
+}
+```
+
+### Generating A Repository Via The CLI
+
+A CLI-generated repository for a model `Note` would look like this:
+
+{% include code-caption.html content="src/repositories/note.repository.ts" %}
+
+```ts
+export class NoteRepository extends DefaultCrudRepository<
+  Note,
+  typeof Note.prototype.id,
+  NoteRelations
+> {
+  constructor(@inject('datasources.db') dataSource: DbDataSource) {
+    super(Note, dataSource);
+  }
+}
+```
+
+### Adjust NoteRepository To Use FindByTitleRepositoryMixin
+
+Import the `FindByTitleRepositoryMixin` mixin class factory function, and adjust
+the declaration of the `NoteRepository` class to extend the class returned from
+the mixin function that takes in the `DefaultCrudRepository` superclass an
+input.
+
+{% include code-caption.html content="src/repositories/note.repository.ts" %}
+
+```ts
+import {FindByTitleRepositoryMixin} from '../mixins/findByTitleRepositoryMixin';
+import {DefaultCrudRepository} from '@loopback/repository';
+import {Note, NoteRelations} from '../models';
+import {DbDataSource} from '../datasources';
+import {inject, Constructor} from '@loopback/core';
+
+export class NoteRepository extends FindByTitleRepositoryMixin<
+  Note,
+  Constructor<
+    DefaultCrudRepository<Note, typeof Note.prototype.id, NoteRelations>
+  >
+>(DefaultCrudRepository) {
+  constructor(@inject('datasources.db') dataSource: DbDataSource) {
+    super(Note, dataSource);
+  }
+}
+```
+
+We have now added the `findByTitle` method to a repository via a mixin class
+factory function.
+
+### Defining A Controller Mixin Class Factory Function
+
+In **src/mixins/findByTitleControllerMixin.ts**, let's define the mixin class
+factory function `FindByTitleControllerMixin` which adds the `findByTitle`
+method to any controller.
+
+{% include code-caption.html content="src/mixins/src/mixins/findByTitleControllerMixin.ts" %}
+
+```ts
+import {Constructor} from '@loopback/context';
+import {Model} from '@loopback/repository';
+import {FindByTitleInterface} from './findByTitleInterface';
+import {param, get, getModelSchemaRef} from '@loopback/rest';
+
+export interface FindByTitleControllerMixinParms {
+  basePath: string;
+  modelClass: Constructor<{}>;
+  modelClassName: string;
+}
+
+export function FindByTitleControllerMixin<
+  M extends Model,
+  T extends Constructor<any>
+>(superClass: T, parms: FindByTitleControllerMixinParms) {
+  class MixedController extends superClass implements FindByTitleInterface<M> {
+    constructor(...args: any[]) {
+      super(...args);
+    }
+
+    @get(parms.basePath + '/findByTitle/{title}', {
+      responses: {
+        '200': {
+          description: `Array of ${parms.modelClassName} model instances`,
```

Comment:
I believe `modelClassName` argument is redundant, you can get the model name via `parms.modelClass.modelName`.

https://github.com/strongloop/loopback-next/blob/abc4c07863060921abf965af3f29718c0caab0b8/packages/repository/src/model.ts#L203-L206

Reply:
Hmmm. It doesn't work

![image](https://user-images.githubusercontent.com/6864736/75309925-739ba780-5820-11ea-9169-33dfb53acac0.png)

Reply:
@bajtos

Reply:
I wish there was a runnable code I could use to try this out myself and tinker with the types.

I think the problem is caused by the fact your `modelClass` argument is specified as `Constructor<object>`, which means any class can be provided as the value and therefore we cannot guarantee that `modelName` static property is available.

IMO, this is a bug, `FindByTitleControllerMixin` should allow only Model constructors. Here is the type definition that should (hopefully) work:

```ts
export interface FindByTitleControllerMixinOptions {
  basePath: string;
  modelClass: typeof Model;
}
```